drv/lpc55-swd/src/main.rs

```diff
@@ -1058,21 +1297,626 @@ impl ServerImpl {
+            Err(e) => {
+                ringbuf_entry!(Trace::AttestError(e));
+                Err(e)
+            }
+        }
+    }
+
+    // Return true if necessary work was done.
+    // Return false if any current SP measurement should be invalidated.
+    fn do_handle_sp_reset(&mut self) -> bool {
+        let start = sys_get_timer().now;
+        let gpio = Pins::from(self.gpio);
+        const SLOT: PintSlot = ROT_TO_SP_RESET_L_IN_PINT_SLOT;
+        let mut need_undo = Undo::from_bits_retain(0);
+
+        // Did SP_RESET transition to Zero?
+        if !gpio
+            .pint_op(SLOT, PintOp::Detected, PintCondition::Falling)
+            .map_or(false, |v| v.unwrap_lite())
+        {
+            // Use of sys_irq_control_clear_pending(...) should avoid
+            // appearance of a "spurious" intrerrupt.
+            // Otherwise, cases where we assert SP_RESET then clean-up the PINT
+            // condition will have a pending notification.
+            ringbuf_entry!(Trace::SpResetNotAsserted);
+            return true; // no work required.
+        }
+
+        // Not ok yet: A reset happened. If we don't get a measurement then
+        // make sure that the old one is invalidated.
+        let mut error = false;
+
+        ringbuf_entry!(Trace::SpResetAsserted);
+
+        // This notification handler should be compatible with watchdog but
+        // will result in the invalidation of any dumps held in the SP if successful.
+
+        // TODO: confirm that bank-flipping SP update watchdog is working.
+
+        self.init = true; // we don't want GPIO pin reconfiguration.
+        if self.do_setup_swd().is_ok() {
+            ringbuf_entry!(Trace::SetupSwdOk);
+            need_undo |= Undo::SWD;
+        } else {
+            // We may have interrupted dumper or watchdog activity.
+            return false; // Cannot make the required measurement.
+        }
+
+        // Armv7-M Arch Ref:
+        // C1.4.1 Entering Debug state on leaving reset state
+        //
+        // To force the processor to enter Debug state as soon as it
+        // comes out of reset, a debugger sets DHCSR.C_DEBUGEN to 1, to
+        // enable Halting debug, and sets DEMCR.VC_CORERESET to 1 to
+        // enable vector catch on the Reset exception. When the
+        // processor comes out of reset it sets DHCSR.C_HALT to 1,
+        // and enters Debug state.
+
+        // If we are late to the SP_RESET party, we're still not that late.
+        // In any case, keep/force the SP into a reset condition.
+        // Though AIRCR::SYSRESETREQ can be used to effect a local reset.
+        // that does not affect the whole SP SoC.
+        // So, use the SP_RESET GPIO.
+        //
+
+        // Setting up to inject the measurement program into the SP
+        // has several potential failures. Use this `prep` closure
+        // and `need_undo` state to keep from indenting too much.
+        let mut prep = |undo: &mut Undo| -> Result<(), ()> {
```

Comment:
Nit: you can capture `need_undo` automatically and don't need to pass it as an argument (so `*undo` in the function body would just become `need_undo`)

Reply:
Done